inferno-core/src/Inferno/Infer.hs

```diff
@@ -1206,20 +1257,128 @@ compose :: Subst -> Subst -> Subst
 (Subst s1) `compose` (Subst s2) = Subst $ Map.map (apply (Subst s1)) s2 `Map.union` s1
 
 -- | Run the constraint solver
-runSolve :: Set.Set TypeClass -> [Constraint] -> Either [TypeError SourcePos] Subst
-runSolve allClasses cs = runIdentity $ runExceptT $ flip runReaderT allClasses $ solver st
+runSolve :: Int -> Set.Set TypeClass -> [Constraint] -> Either [TypeError SourcePos] Subst
+runSolve varCount allClasses cs = runIdentity $ runExceptT $ flip runReaderT allClasses $ solver varCount st
   where
     st = (emptySubst, cs)
 
-unifyMany :: [TypeError SourcePos] -> [InfernoType] -> [InfernoType] -> Solve Subst
+unifyMany :: [TypeError SourcePos] -> [InfernoType] -> [InfernoType] -> SolveState Int Subst
```

Comment:
Once everything works, I can clean this up and use some `FreshVarState` or something instead of `Int`